File: python/92_Reverse_Linked_List_II.py
# coding=utf-8
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None
class Solution:
    # @param {ListNode} head
    # @param {integer} m
    # @param {integer} n
    # @return {ListNode}
    def reverseBetween(self, head, m, n):
        if not m or not n or m>n:
            return head
        dummy=ListNode(0)
        dummy.next=head
        cur=dummy.__next__
        i=1
        start=dummy
        while i<m and cur:
            start=cur
            cur=cur.__next__
            i+=1
        if cur:
            stop=cur
            runner=cur.__next__
            while i<n and runner:
                p=runner.__next__
                runner.next=cur
                cur=runner
                runner=p
                i+=1
            stop.next=runner
            start.next=cur
        return dummy.__next__
        
def printList(head):
    if not head:
        print("None")
    while head.__next__:
        print(head.val,"->", end=' ')
        head=head.__next__
    print(head.val)
    
so=Solution()
h=head=ListNode(1)
for i in range(2,6):
    h.next=ListNode(i)
    h=h.__next__
head=so.reverseBetween(head, -1, 3)
printList(head)
head=so.reverseBetween(head, 2, 2)
printList(head)

File: Java/92_Reverse_Linked_List_II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
// 构造出了一个leetcode本身的解会出错的testcase：
// [1,2,3] 1 7
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if(n<=0||m>=n||head==null||head.next==null)return head;
        ListNode dummy = new ListNode(0) ,runner;
        dummy.next = head;
        n=n-m;// 要操作的节点数，如果下面的条件是>0的话这里就应该+1
        runner = dummy;
        while(runner.next!=null && m>1){// 注意节点从1开始计数
            runner = runner.next;
            m--;
        }
        ListNode np = runner.next, p;
        runner.next=null;
        ListNode tail = np;// 保存翻转之后的尾节点，好接上剩余的链表
        while(np!=null&&n>=0){
            p=np.next;
            np.next=runner.next;
            runner.next=np;
            np=p;
            n--;
        }
        tail.next=np;
        return dummy.next;
    }
}